src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/RestoreService.java

```diff
@@ -0,0 +1,129 @@
+package org.niis.xroad.restapi.service;
+
+import ee.ria.xroad.common.identifier.SecurityServerId;
+
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.niis.xroad.restapi.cache.CurrentSecurityServerId;
+import org.niis.xroad.restapi.exceptions.DeviationAwareRuntimeException;
+import org.niis.xroad.restapi.repository.BackupRepository;
+import org.niis.xroad.restapi.util.FormatUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Service;
+
+import java.io.File;
+
+/**
+ * service class for restoring security server configuration from a backup
+ */
+@Slf4j
+@Service
+@PreAuthorize("isAuthenticated()")
+public class RestoreService {
+    @Setter
+    private String configurationRestoreScriptPath;
+    @Setter
+    private String configurationRestoreScriptArgs;
+
+    private final ExternalProcessRunner externalProcessRunner;
+    private final CurrentSecurityServerId currentSecurityServerId;
+    private final BackupRepository backupRepository;
+    private final NotificationService notificationService;
+
+    @Autowired
+    public RestoreService(ExternalProcessRunner externalProcessRunner,
+            @Value("${script.restore-configuration.path}") String configurationRestoreScriptPath,
+            @Value("${script.restore-configuration.args}") String configurationRestoreScriptArgs,
+            CurrentSecurityServerId currentSecurityServerId, BackupRepository backupRepository,
+            NotificationService notificationService) {
+        this.externalProcessRunner = externalProcessRunner;
+        this.configurationRestoreScriptPath = configurationRestoreScriptPath;
+        this.configurationRestoreScriptArgs = configurationRestoreScriptArgs;
+        this.currentSecurityServerId = currentSecurityServerId;
+        this.backupRepository = backupRepository;
+        this.notificationService = notificationService;
+    }
+
+    /**
+     * Restores the security server configuration from a backup. Any tokens that are not software tokens are logged
+     * out by the current restore script.
+     * @param fileName name of the backup file
+     * @throws BackupFileNotFoundException
+     * @throws InterruptedException execution of the restore script was interrupted
+     * @throws RestoreInProgressException if restore is already in progress
+     */
+    public synchronized void restoreFromBackup(String fileName) throws BackupFileNotFoundException,
+            InterruptedException, RestoreInProgressException {
+        if (notificationService.getBackupRestoreRunningSince() != null) {
+            // should not happen because the method is synchronized
+            throw new RestoreInProgressException("There is a restore (started at "
+                    + notificationService.getBackupRestoreRunningSince() + ") already in progress");
+        }
```

Comment:
I had kind of a difficult choice between options:
1. mark the method as `synchronized` to prevent concurrency
2. make it throw an exception if trying to run concurrently

I ended up adding both.

Throwing an exception makes it possible to run the restore concurrently (in theory) if multiple users trigger it simultaneously. The user is already seeing an alert saying that there is a restore in progress so getting an error would feel natural if the user tried to trigger another restore.

In it's current state the synchronized nature of the method overrides the `notificationService.getBackupRestoreRunningSince()` check making it redundant.

Reply:
I think synchronized is a must, since we do not want two threads getting past the `notificationService.getBackupRestoreRunningSince` check concurrently.

Since method is synchronized we can reason that notificationService.getBackupRestoreRunningSince check is unnecessary, since we "know" it cannot happen. I don't think it is bad to have it, as kind of defensive programming, and since there's not too much complexity cost for it. 

But since it is just defensive programming, and should not happen, I think it should not be a checked exception. Sometimes it's said those kind of checks should be done with asserts but a runtime exception is fine too. We don't want to complicate the API and burden the method's client with a checked exception type which is impossible, IMO. As we want to limit the number of exceptions (and I also had a different checked exception type in mind that I wanted to add...).

Reply:
Sound like good reasoning. I changed the checked exception into a runtime exception instead.